File: neetcode/Subsets.py
from typing import List

# M-1
class Solution:
    def __init__(self):
        self.ans=[]

    def recursion(self, rec:List[int], a:List[int]):
        if (len(rec)==0):
            self.ans.append(a[:])
            return
        
        self.recursion(rec[1:],a)
        e = rec[0]
        a.append(e)
        self.recursion(rec[1:],a)
        a.pop()

    def subsets(self, nums: List[int]) -> List[List[int]]:
        # for i in range(len(nums)):
        #     for j in range(len(nums)):
        a=[]
        self.recursion(nums,a)
        return self.ans

obj = Solution()
print(obj.subsets([1,2,3,4]))

# M-2

File: neetcode/Subsets.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<vector<int > > ans;
    void r(vector<int> nums,vector<int>& a) {
        if(nums.size()==0){
            ans.push_back(a);
            return;
        }
        int temp=nums[0];
        nums.erase(nums.begin());
        r(nums,a);
        a.push_back(temp);
        r(nums,a);
        a.pop_back();
    }

    vector<vector<int> > subsets(vector<int>& nums) {
        vector<int>a;
        r(nums,a);
        return ans;
    }
};
int main (){
    Solution obj;
    vector<int> nums;
    nums.push_back(2);
    nums.push_back(3);
    nums.push_back(1);
    obj.subsets(nums);
    return 0;
}